---
title: 작업 취소
description: .NET에서 취소 토큰을 사용하여 Task 및 Task<TResult> 클래스에서 지원되는 작업 취소를 이해합니다.
ms.date: 03/30/2017
dev_langs:
- csharp
- vb
helpviewer_keywords:
- tasks, cancellation
- asynchronous task cancellation
ms.assetid: 3ecf1ea9-e399-4a6a-a0d6-8475f48dcb28
ms.openlocfilehash: deea3eaaa1e652d44a71e953975d776d4898a9b3
ms.sourcegitcommit: 965a5af7918acb0a3fd3baf342e15d511ef75188
ms.translationtype: HT
ms.contentlocale: ko-KR
ms.lasthandoff: 11/18/2020
ms.locfileid: "94830014"
---
# <a name="task-cancellation"></a><span data-ttu-id="3ad53-103">작업 취소</span><span class="sxs-lookup"><span data-stu-id="3ad53-103">Task cancellation</span></span>

<span data-ttu-id="3ad53-104"><xref:System.Threading.Tasks.Task?displayProperty=nameWithType> 및 <xref:System.Threading.Tasks.Task%601?displayProperty=nameWithType> 클래스는 취소 토큰을 사용하는 방법으로 취소 기능을 지원합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-104">The <xref:System.Threading.Tasks.Task?displayProperty=nameWithType> and <xref:System.Threading.Tasks.Task%601?displayProperty=nameWithType> classes support cancellation through the use of cancellation tokens.</span></span> <span data-ttu-id="3ad53-105">자세한 내용은 [관리되는 스레드의 취소](../threading/cancellation-in-managed-threads.md)를 참조하세요.</span><span class="sxs-lookup"><span data-stu-id="3ad53-105">For more information, see [Cancellation in Managed Threads](../threading/cancellation-in-managed-threads.md).</span></span> <span data-ttu-id="3ad53-106">작업 클래스에서 취소하려면 취소할 수 있는 작업을 나타내는 사용자 대리자와 취소를 요청한 코드 간의 협조가 필요합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-106">In the Task classes, cancellation involves cooperation between the user delegate, which represents a cancelable operation, and the code that requested the cancellation.</span></span> <span data-ttu-id="3ad53-107">성공적으로 취소하려면 <xref:System.Threading.CancellationTokenSource.Cancel%2A?displayProperty=nameWithType> 메서드를 호출하는 요청 코드와 적절한 시간에 작업을 종료하는 사용자 대리자가 필요합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-107">A successful cancellation involves the requesting code calling the <xref:System.Threading.CancellationTokenSource.Cancel%2A?displayProperty=nameWithType> method and the user delegate terminating the operation in a timely manner.</span></span> <span data-ttu-id="3ad53-108">다음 방법 중 하나를 사용하여 작업을 종료할 수 있습니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-108">You can terminate the operation by using one of these options:</span></span>  
  
- <span data-ttu-id="3ad53-109">단순히 대리자에서 반환합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-109">By simply returning from the delegate.</span></span> <span data-ttu-id="3ad53-110">대부분의 경우에는 이 방법으로 충분하지만 이 방법으로 취소된 작업 인스턴스는 <xref:System.Threading.Tasks.TaskStatus.RanToCompletion?displayProperty=nameWithType> 상태가 아니라 <xref:System.Threading.Tasks.TaskStatus.Canceled?displayProperty=nameWithType> 상태로 전환됩니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-110">In many scenarios this is sufficient; however, a task instance that is canceled in this way transitions to the <xref:System.Threading.Tasks.TaskStatus.RanToCompletion?displayProperty=nameWithType> state, not to the <xref:System.Threading.Tasks.TaskStatus.Canceled?displayProperty=nameWithType> state.</span></span>  
  
- <span data-ttu-id="3ad53-111"><xref:System.OperationCanceledException> 을 throw하고 취소가 요청된 토큰을 전달합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-111">By throwing a <xref:System.OperationCanceledException> and passing it the token on which cancellation was requested.</span></span> <span data-ttu-id="3ad53-112">이렇게 하려면 <xref:System.Threading.CancellationToken.ThrowIfCancellationRequested%2A> 메서드를 사용하는 것이 좋습니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-112">The preferred way to do this is to use the <xref:System.Threading.CancellationToken.ThrowIfCancellationRequested%2A> method.</span></span> <span data-ttu-id="3ad53-113">이 방법으로 취소된 작업은 Canceled 상태로 전환되므로 호출 코드에서는 이 상태를 통해 작업이 취소 요청에 응답했음을 확인할 수 있습니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-113">A task that is canceled in this way transitions to the Canceled state, which the calling code can use to verify that the task responded to its cancellation request.</span></span>  
  
 <span data-ttu-id="3ad53-114">다음 예제에서는 예외를 throw하는 작업 취소의 기본적인 패턴을 보여 줍니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-114">The following example shows the basic pattern for task cancellation that throws the exception.</span></span> <span data-ttu-id="3ad53-115">토큰은 사용자 대리자와 작업 인스턴스 자체에 전달됩니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-115">Note that the token is passed to the user delegate and to the task instance itself.</span></span>  
  
 [!code-csharp[TPL_Cancellation#02](../../../samples/snippets/csharp/VS_Snippets_Misc/tpl_cancellation/cs/snippet02.cs#02)]
 [!code-vb[TPL_Cancellation#02](../../../samples/snippets/visualbasic/VS_Snippets_Misc/tpl_cancellation/vb/module1.vb#02)]  
  
 <span data-ttu-id="3ad53-116">자세한 예제는 [방법: 작업 및 해당 자식 취소](how-to-cancel-a-task-and-its-children.md)를 참조하세요.</span><span class="sxs-lookup"><span data-stu-id="3ad53-116">For a more complete example, see [How to: Cancel a Task and Its Children](how-to-cancel-a-task-and-its-children.md).</span></span>  
  
 <span data-ttu-id="3ad53-117">작업 인스턴스는 사용자 코드에서 throw된 <xref:System.OperationCanceledException> 이 관찰될 경우 예외의 토큰을 관련 토큰, 즉 해당 작업을 만든 API에 전달된 토큰과 비교합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-117">When a task instance observes an <xref:System.OperationCanceledException> thrown by user code, it compares the exception's token to its associated token (the one that was passed to the API that created the Task).</span></span> <span data-ttu-id="3ad53-118">두 토큰이 동일하고 토큰의 <xref:System.Threading.CancellationToken.IsCancellationRequested%2A> 속성이 true를 반환하면 해당 작업은 이를 취소 승인으로 해석하고 Canceled 상태로 전환됩니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-118">If they are the same and the token's <xref:System.Threading.CancellationToken.IsCancellationRequested%2A> property returns true, the task interprets this as acknowledging cancellation and transitions to the Canceled state.</span></span> <span data-ttu-id="3ad53-119"><xref:System.Threading.Tasks.Task.Wait%2A> 또는 <xref:System.Threading.Tasks.Task.WaitAll%2A> 메서드를 사용하여 작업이 완료될 때까지 대기하지 않는 경우 해당 작업은 단순히 상태를 <xref:System.Threading.Tasks.TaskStatus.Canceled>로 설정합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-119">If you do not use a <xref:System.Threading.Tasks.Task.Wait%2A> or <xref:System.Threading.Tasks.Task.WaitAll%2A> method to wait for the task, then the task just sets its status to <xref:System.Threading.Tasks.TaskStatus.Canceled>.</span></span>  
  
 <span data-ttu-id="3ad53-120">Canceled 상태로 전환되는 작업에서 대기 중인 경우에는 <xref:System.Threading.Tasks.TaskCanceledException?displayProperty=nameWithType> 예외가 <xref:System.AggregateException> 예외에 래핑된 상태로 throw됩니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-120">If you are waiting on a Task that transitions to the Canceled state, a <xref:System.Threading.Tasks.TaskCanceledException?displayProperty=nameWithType> exception (wrapped in an <xref:System.AggregateException> exception) is thrown.</span></span> <span data-ttu-id="3ad53-121">이 예외는 오류가 아니라 성공적인 취소를 나타냅니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-121">Note that this exception indicates successful cancellation instead of a faulty situation.</span></span> <span data-ttu-id="3ad53-122">따라서 작업의 <xref:System.Threading.Tasks.Task.Exception%2A> 속성은 `null`을 반환합니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-122">Therefore, the task's <xref:System.Threading.Tasks.Task.Exception%2A> property returns `null`.</span></span>  
  
 <span data-ttu-id="3ad53-123">토큰의 <xref:System.Threading.CancellationToken.IsCancellationRequested%2A> 속성이 false를 반환하거나 예외의 토큰이 작업의 토큰과 일치하지 않는 경우 <xref:System.OperationCanceledException> 은 일반적인 예외처럼 처리되므로 작업이 Faulted 상태로 전환됩니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-123">If the token's <xref:System.Threading.CancellationToken.IsCancellationRequested%2A> property returns false or if the exception's token does not match the Task's token, the <xref:System.OperationCanceledException> is treated like a normal exception, causing the Task to transition to the Faulted state.</span></span> <span data-ttu-id="3ad53-124">또한 다른 예외가 있는 경우에도 작업이 Faulted 상태로 전환됩니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-124">Also note that the presence of other exceptions will also cause the Task to transition to the Faulted state.</span></span> <span data-ttu-id="3ad53-125">완료된 작업의 상태는 <xref:System.Threading.Tasks.Task.Status%2A> 속성에서 가져올 수 있습니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-125">You can get the status of the completed task in the <xref:System.Threading.Tasks.Task.Status%2A> property.</span></span>  
  
 <span data-ttu-id="3ad53-126">취소를 요청한 후에도 작업에서 일부 항목의 처리를 계속 진행할 수 있습니다.</span><span class="sxs-lookup"><span data-stu-id="3ad53-126">It is possible that a task may continue to process some items after cancellation is requested.</span></span>  
  
## <a name="see-also"></a><span data-ttu-id="3ad53-127">참조</span><span class="sxs-lookup"><span data-stu-id="3ad53-127">See also</span></span>

- [<span data-ttu-id="3ad53-128">관리되는 스레드의 취소</span><span class="sxs-lookup"><span data-stu-id="3ad53-128">Cancellation in Managed Threads</span></span>](../threading/cancellation-in-managed-threads.md)
- [<span data-ttu-id="3ad53-129">방법: 작업 및 해당 자식 취소</span><span class="sxs-lookup"><span data-stu-id="3ad53-129">How to: Cancel a Task and Its Children</span></span>](how-to-cancel-a-task-and-its-children.md)
